runners/core-java/src/main/java/org/apache/beam/runners/core/SplittableParDo.java

```diff
@@ -0,0 +1,403 @@
+      // producing a limited amount of output.
+      DoFn.ProcessContinuation cont =
+          invoker.invokeProcessElement(
+              makeContext(c, element, tracker, residual, internals), wrapTracker(tracker));
+
+      if (cont == null) {
+        // All work for this element/restriction is completed. Clear state and release hold.
+        elementState.clear();
+        restrictionState.clear();
+        holdState.clear();
+        return;
+      }
+      if (residual[0] == null) {
+        // This means the call completed unsolicited, and the context produced by makeContext()
+        // did not take a checkpoint. Take one now.
+        residual[0] = checkNotNull(tracker.checkpoint());
+      }
+
+      // Save state for resuming.
+      if (wasFirstCall) {
+        elementState.write(element);
+      }
+      restrictionState.write(residual[0]);
+      if (cont.getFutureOutputWatermark() != null) {
+        holdState.add(cont.getFutureOutputWatermark());
+      }
+      // Set a timer to continue processing this element.
+      timerInternals.setTimer(
+          TimerInternals.TimerData.of(
+              StateNamespaces.global(),
+              Instant.now().plus(cont.getResumeDelay()),
+              TimeDomain.PROCESSING_TIME));
+    }
+
+    private DoFn<InputT, OutputT>.ProcessContext makeContext(
+        final ProcessContext baseContext,
+        final WindowedValue<InputT> element,
+        final TrackerT tracker,
+        final RestrictionT[] residualRestrictionHolder,
+        final WindowingInternals<KeyedWorkItem<String, KV<InputT, RestrictionT>>, OutputT>
+            internals) {
+      return fn.new ProcessContext() {
+        // Commit at least once every 10k output records.  This keeps the watermark advancing
+        // smoothly, and ensures that not too much work will have to be reprocessed in the event of
+        // a crash.
+        // TODO: Also commit at least once every N seconds (runner-specific parameter).
+        private static final int MAX_OUTPUTS_PER_BUNDLE = 10000;
+
+        private int numOutputs = 0;
+
+        public InputT element() {
+          return element.getValue();
+        }
+
+        public Instant timestamp() {
+          return element.getTimestamp();
+        }
+
+        public PaneInfo pane() {
+          return element.getPane();
+        }
+
+        public void output(OutputT output) {
+          internals.outputWindowedValue(
+              output, element.getTimestamp(), element.getWindows(), element.getPane());
+          noteOutput();
+        }
+
+        public void outputWithTimestamp(OutputT output, Instant timestamp) {
+          internals.outputWindowedValue(output, timestamp, element.getWindows(), element.getPane());
+          noteOutput();
+        }
+
+        private void noteOutput() {
+          if (++numOutputs > MAX_OUTPUTS_PER_BUNDLE) {
+            // Request a checkpoint. The fn *may* produce more output, but hopefully not too much.
```

Comment:
I think I meant tracking how much more than the requested split was produced. It can definitely be future work. But, for an arbitrary DoFn, it would be good to know that if we asked for a checkpoint at 10000 elements, if we got 10000, 10001, or 500000 (one of these is less finely splittable than the others).